test/tools/yulopti.cpp

```diff
@@ -87,37 +89,70 @@ class YulOpti
 		}.printErrorInformation(_errors);
 	}
 
-	void parse(string const& _input)
+	shared_ptr<Object> getSubObject(shared_ptr<Object> const& _object, string const& _qualifiedPath)
+	{
+		if (_qualifiedPath.empty() || _qualifiedPath == _object->name.str())
+			return _object;
+
+		yulAssert(
+			boost::algorithm::starts_with(_qualifiedPath, _object->name.str() + "."),
+			"Assembly object not found."
+		);
+
+		string const subObjectPath = _qualifiedPath.substr(_object->name.str().length() + 1);
+		string const subObjectName = subObjectPath.substr(0, subObjectPath.find_first_of('.'));
+
+		auto subObjectIt = ranges::find_if(
+			_object->subObjects,
+			[&subObjectName](auto const& _subObject) { return _subObject->name.str() == subObjectName; }
+		);
+
+		yulAssert(
+			subObjectIt != _object->subObjects.end(),
+			"Assembly object not found."
+		);
+
+		auto subObject = dynamic_pointer_cast<Object>(*subObjectIt);
+
+		yulAssert(
+			subObject != nullptr,
+			"Assembly object may not contain code."
+		);
+
+		return getSubObject(subObject, subObjectPath);
+	}
+
+	void parse(string const& _input, string const& _objectPath)
 	{
 		ErrorList errors;
 		ErrorReporter errorReporter(errors);
-		CharStream _charStream(_input, "");
+		CharStream charStream(_input, "");
+
 		try
 		{
-			m_ast = yul::Parser(errorReporter, m_dialect).parse(_charStream);
-			if (!m_ast || !errorReporter.errors().empty())
+			ObjectParser parser(errorReporter, m_dialect);
+
+			auto scanner = make_shared<Scanner>(charStream);
+
+			if (!m_inputWasCodeBlock && scanner->currentToken() == Token::LBrace)
+				m_inputWasCodeBlock = true;
+
+			shared_ptr<Object> object = parser.parse(scanner, false);
+
+			if (!object || !errorReporter.errors().empty())
 			{
 				cerr << "Error parsing source." << endl;
-				printErrors(_charStream, errors);
-				throw std::runtime_error("Could not parse source.");
-			}
-			m_analysisInfo = make_unique<yul::AsmAnalysisInfo>();
-			AsmAnalyzer analyzer(
-				*m_analysisInfo,
-				errorReporter,
-				m_dialect
-			);
-			if (!analyzer.analyze(*m_ast) || !errorReporter.errors().empty())
-			{
-				cerr << "Error analyzing source." << endl;
-				printErrors(_charStream, errors);
-				throw std::runtime_error("Could not analyze source.");
+				printErrors(charStream, errors);
+				throw runtime_error("Could not parse source.");
 			}
+
+			m_object = getSubObject(object, _objectPath);
```

Comment:
Don't we have to keep the whole object? If you do this, I'd expect `parse()` to fail if called multiple times (unless the nested object happens to have the same path inside it - but even then it would be a different subobject than requested).

Reply:
Ah, I see we don't. We only run `parse()` once. So I guess this is fine as long as a nested object cannot refer to any objects on upper levels and I think that currently it can't.